test/serve-asset-middleware-test.js

```diff
@@ -0,0 +1,206 @@
+      });
+
+      var middleware = watcherMiddleware(watcher, {
+        autoIndex: false
+      });
+      var serveMiddleware = serveAssetMiddleware();
+
+      server = new TestHTTPServer(middleware);
+      server.addMiddleware(serveMiddleware);
+
+      return server.start()
+        .then(function(info) {
+          return server.request('/index.html', {
+            info: info
+          });
+        })
+        .then(function(content) {
+          expect(content).to.match(/This is broccoli middleware page/);
+        });
+    });
+
+    it('no response headers are set & it calls the next middleware when file is not found', function(done) {
+      var watcher = RSVP.Promise.resolve({
+        'directory': fixture('basic-file')
+      });
+
+      var middleware = watcherMiddleware(watcher, {
+        autoIndex: false
+      });
+
+      var serveAssetMiddlewareFn = serveAssetMiddleware();
+      var wrapperMiddleware = function(req, resp/*, next*/) {
+        serveAssetMiddlewareFn(req, resp, function() {
+          // assert response headers
+          expect(resp.getHeader('Cache-Control')).to.be.equal('private, max-age=0, must-revalidate');
+          expect(resp.getHeader('Last-Modified')).to.not.be.ok;
+          expect(resp.getHeader('Content-Type')).to.not.be.ok;
+          done();
+        });
+      }
+
+
+      server = new TestHTTPServer(middleware);
+      server.addMiddleware(wrapperMiddleware);
+
+      server.start()
+        .then(function(info) {
+          return server.request('/non-existent-file', {
+            info: info
+          });
+        });
+    });
+
+    it('bypasses middleware if request is a directory and autoIndex is set to false', function(done) {
+      var watcher = RSVP.Promise.resolve({
+        'directory': fixture('no-index')
+      });
+
+      var middleware = watcherMiddleware(watcher, {
+        autoIndex: false
+      });
+
+      var serveAssetMiddlewareFn = serveAssetMiddleware();
+      var wrapperMiddleware = function(req, resp /*next*/) {
+        serveAssetMiddlewareFn(req, resp, function() {
+          var broccoliHeader = req.headers['x-broccoli'];
+          expect(broccoliHeader.filename).to.equal('');
+          done();
+        });
+      };
+
+      server = new TestHTTPServer(middleware);
+      server.addMiddleware(wrapperMiddleware);
+
+      server.start()
+        .then(function(info) {
+          return server.request('', {
+            info: info
+          });
+        });
+    });
+
+    it('responds with directory structure template if request is a directory and autoIndex is set to true', function() {
+      var watcher = RSVP.Promise.resolve({
+        'directory': fixture('no-index')
+      });
+
+      var middleware = watcherMiddleware(watcher, {
+        autoIndex: true
+      });
+
+      var serveMiddleware = serveAssetMiddleware();
+
+      server = new TestHTTPServer(middleware);
+      server.addMiddleware(serveMiddleware);
```

Comment:
The wrapperMiddleware creates a closure to allow doing assertions. I could otherwise just create a another middleware to do the assertions after serveAssetMiddleware is finished executing.